modules/dists/BlockDist.chpl

```diff
@@ -1765,10 +1765,11 @@ proc BlockArr.doiScan(op, dom) where (rank == 1) &&
       // local array, local domain, and local result elements
       ref myLocArrDesc = locArr[locid];
       ref myLocArr = myLocArrDesc.myElems;
+      ref myLocRes = res._value.locArr[locid].myElems;
       const ref myLocDom = myLocArr.domain;
 
       // Compute the local pre-scan on our local array
-      var (numTasks, rngs, state, tot) = myLocArr._value.chpl__preScan(myop, res, myLocDom[dom]);
+      var (numTasks, rngs, state, tot) = myLocArr._value.chpl__preScan(myop, myLocRes, myLocDom[dom]);
```

Comment:
Did you happen to look into whether we could/should just remove the domain from the preScan() interface?  It seems likely that we're now in a situation where we're always passing in the same domain that defines the array we're passing in?  (Was this the thing I took a TODO to look into a few weeks back, but apparently never got to?).

Reply:
I'm not sure. That was added as part of https://github.com/chapel-lang/chapel/pull/12820.

I think the TODO was to see if the work in this PR made sense. i.e. are we always operating on the local portion of the result array? I don't see how we wouldn't be, but then I wasn't sure what happens with slices or something.

Reply:
Ah, OK.  I think that if the preScan method was defined on array records rather than classes and we queried its domain from that, we could drop it; but since it acts on array classes, the logical domains that control them may be distinct from their logical ones (since they describe the allocation).  So I agree with leaving it as-is.

With respect to the locality of the result array...  I think the killer case (if there is one) would be something like this:

```chapel
var A = newBlockArr({1..n}, real);
var Aview = A[{1..10}];   // Aview's domain is a DR {1..10}, not a Block array
var B = + scan Aview;
```

So let's see... thinking about what would happen here:

* the compiler would call _array.scan()
* it would call BlockArr.doiScan() passing in the DR `{1..10}` as `dom`, I think
* `res` would be declared over `dom`, so would also have/be a DR domain

Ah, so that suggests this PR wouldn't work because your change is accessing `res` as though it was `BlockDist`.  You should double-check me, though, b/c I don't have a clear workspace to do it in a.t.m.  (and of course, having typed that, I'm also wondering whether we might have other assumptions on `main` that would already make this case not work, or something like that).

Assuming that's correct, I think you could still do a special-case like this, but would need to predicate it on something like `dom` and `res` have the same distribution or aligned distributions or something like that (?).

Reply:
Thanks, these are the types of insight I was looking for since I don't fully understand how this is all threaded through.

Your case as coded works on main and with this PR, but I think you wanted to use `ref Aview`, which fails on both main and this PR because `dom.dist.targetLocDom` isn't defined for DR domains:

```chpl
use BlockDist;
config const n = 20;

var A = newBlockArr({1..n}, real); A = 1;
ref Aview = A[{1..10}];
var B = + scan Aview;

writeln(B.type:string);
writeln(B);
```

```
$ chpl test.chpl
$CHPL_HOME/modules/dists/BlockDist.chpl:1744: In method 'doiScan':
$CHPL_HOME/modules/dists/BlockDist.chpl:1760: error: unresolved call 'DefaultDist.targetLocDom'
```

Reply:
Oops, yes, you're right about the intent to use `ref`.

Reply:
Updated to check that the distribution of the input array and result array are the same in order to trigger this optimization. This reuses the fast follower checks in order to determine this. Note that this will lead to some code bloat since we'll have a runtime check for the distribution, which means we'll generate both the optimized and non-optimized pre/postScan routines.

@e-kayrakli or @benharsh -- I think you're familiar with the fast follower checks. Could you sanity check that I got the right args for those (in particular passing `_value` to the static check and not for the dynamic one.)